docs/language.rst

```diff
@@ -864,3 +864,76 @@ Plug-ins
 For more complex operations, python plugins can be used. Plugins are exposed in the Inmanta language as function calls, such as the template function call. A template
 accepts parameters and returns a value that it computed out of the variables. Each module that is included can also provide plug-ins. These plug-ins are accessible within the namespace of the
 module. The :ref:`module-plugins` section of the module guide provides more details about how to write a plugin.
+
+
+.. _language_unknowns:
+Unknowns
+========
+
+Wherever the configuration model interacts with the outside world (e.g. to fetch external values) :term:`unknown` values
+may be present. These unknowns are propagated through the model to finally end up in the resources that require these unknown
+values. This section describes how unknown values flow through the model, and perhaps equally importantly, where they do not
+flow at all.
+
+.. note::
+    Unknowns are a subtle concept. Luckily, for the majority of model development you don't really need to take them into
+    account. However, for some advanced scenarios it may be important to know how and where they may occur.
+
+For the most part, unknowns are simply propagated along the data flow: statements like assignment statements, constructors
+and lists simply include the unknown in their result like they would any other value. Any expression that can not produce
+a definite result without knowing the value, will return another unknown. And finally, statements that expand the model
+with new blocks based on some value, like the if statement and the for loop, simply do not expand the model with their
+respective blocks for unknowns.
+
+The model below presents some examples of how an unknown propagates.
+
+.. code-block:: inmanta
+
+    # std::env returns an unknown if the environment variable is not (yet) set
+    my_unknown = std::get_env("THIS_ENV_VAR_DOES_NOT_EXIST")
+
+    a = my_unknown  # a is unknown
+    b = [1, 2, my_unknown, 3]  # b is a list with 1 unknown element
+    c = my_unknown is defined  # we can not know if c is null, so c is also unknown
+    d = true or my_unknown  # trivial, value of my_unknown is irrelevant -> d is true
+    e = my_unknown or true  # lazy boolean operator can not compute result without knowing the value -> e is unknown
+    f = (e == my_unknown)  # both e and my_unknown are unknown but they aren't necessarily the same value -> f is unknown
+
+    if my_unknown:
+        # this block is never executed
+    else:
+        # neither is this one
+    end
+
+    for x in my_unknown:
+        # neither is this one
+    end
+
+    for x in [1, 2, my_unknown]:
+        # this block is executed twice: x=1 and x=2
+    end
+
+    g = my_unknown ? true : false  # condition is unknown -> neither branch is executed, result is unknown
+    h = [1 for x in [1, 2, my_unknown]]  # the expression `1` is executed once with x=1 and once with x=2. Unknown is propagated as is -> h = [1, 1, unknown]
```

Comment:
I find these two to be hard to reconcile with the for loop. For me, a list comprehension is 

```
result = [a for item in c if d]
```
with a and d expressions referring to item and c an expression
```
collector = ...
for item in c:
  if d:
    collector.collection += a
result = collector.collection
``` 

but again, I raised this before.